src/gui/widgets/typed_debugger.cc

```diff
@@ -0,0 +1,967 @@
+}
+
+// Returns the appropriate pointer if the given address is in RAM or in the scratchpad, in which case the memory base
+// address is stored in outMemBase; otherwise, a null pointer is returned and no value is assigned to MemBase.
+static uint8_t* getMemoryPointerFor(uint32_t address, uint32_t& outMemBase) {
+    if (isInRAM(address)) {
+        outMemBase = 0x80000000;
+        return PCSX::g_emulator->m_mem->m_psxM;
+    }
+
+    const uint32_t memBase = 0x1f800000;
+    const uint32_t memSize = 1024;
+    if (address >= memBase && address < memBase + memSize) {
+        outMemBase = memBase;
+        return PCSX::g_emulator->m_mem->m_psxH;
+    }
+
+    return nullptr;
+}
+
+static bool equals(const char* lhs, const char* rhs) { return strcmp(lhs, rhs) == 0; }
+
+static bool isPrimitive(const char* type) {
+    // Don't test against "char" since char* is handled specially.
+    return equals(type, "uchar") || equals(type, "u_char") || equals(type, "short") || equals(type, "ushort") ||
+           equals(type, "u_short") || equals(type, "int") || equals(type, "long") || equals(type, "uint") ||
+           equals(type, "u_int") || equals(type, "ulong") || equals(type, "u_long");
+}
+
+void PCSX::Widgets::TypedDebugger::displayBreakpointOptions(WatchTreeNode* node, const uint32_t address,
+                                                            uint8_t* memData, const uint32_t memBase) {
+    const auto readBreakpointButtonName = fmt::format(f_("Add read breakpoint##{}{}"), node->name.c_str(), address);
+    if (ImGui::Button(readBreakpointButtonName.c_str())) {
+        auto* newBreakpoint = PCSX::g_emulator->m_debug->addBreakpoint(address, PCSX::Debug::BreakpointType::Read,
+                                                                       node->size, _("Typed Debugger"));
+        m_watchBreakpoints.push_back(newBreakpoint);
+    }
+    ImGui::SameLine();
+    const auto writeBreakpointButtonName = fmt::format(f_("Add write breakpoint##{}{}"), node->name.c_str(), address);
+    if (ImGui::Button(writeBreakpointButtonName.c_str())) {
+        auto* newBreakpoint = PCSX::g_emulator->m_debug->addBreakpoint(address, PCSX::Debug::BreakpointType::Write,
+                                                                       node->size, _("Typed Debugger"));
+        m_watchBreakpoints.push_back(newBreakpoint);
+    }
+    ImGui::SameLine();
+    const auto logReadsWritesButtonName = fmt::format(f_("Log reads and writes##{}{}"), node->name.c_str(), address);
+    if (ImGui::Button(logReadsWritesButtonName.c_str())) {
+        PCSX::Debug::BreakpointInvoker logReadsWritesInvoker =
+            [this, node](const PCSX::Debug::Breakpoint* self, uint32_t address, unsigned width, const char* cause) {
+                if (!node) {
+                    return false;
+                }
+
+                const auto pc = g_emulator->m_cpu->m_regs.pc;
+                std::string funcName = getFunctionNameFromInstructionAddress(pc);
```

Comment:
If changing the return type of this function, don't forget to also change these types.